autonat.go

```diff
@@ -75,198 +70,227 @@ func NewAutoNAT(ctx context.Context, h host.Host, getAddrs GetAddrs) AutoNAT {
+
+	emitReachabilityChanged, _ := h.EventBus().Emitter(new(event.EvtLocalReachabilityChanged), eventbus.Stateful)
 
 	as := &AmbientAutoNAT{
-		ctx:      ctx,
-		host:     h,
-		getAddrs: getAddrs,
-		peers:    make(map[peer.ID][]ma.Multiaddr),
-		status:   NATStatusUnknown,
-
-		emitUnknown: emitUnknown,
-		emitPublic:  emitPublic,
-		emitPrivate: emitPrivate,
+		ctx:            ctx,
+		host:           h,
+		getAddrs:       getAddrs,
+		candidatePeers: make(chan network.Conn, 5),
+		observations:   make(chan autoNATResult, 1),
+
+		subAddrUpdated: subAddrUpdated,
+
+		emitReachabilityChanged: emitReachabilityChanged,
 	}
+	as.status.Store(autoNATResult{network.ReachabilityUnknown, nil})
 
 	h.Network().Notify(as)
 	go as.background()
 
 	return as
 }
 
-func (as *AmbientAutoNAT) Status() NATStatus {
-	as.mx.Lock()
-	defer as.mx.Unlock()
-	return as.status
+// Status returns the AutoNAT observed reachability status.
+func (as *AmbientAutoNAT) Status() network.Reachability {
+	s := as.status.Load().(autoNATResult)
+	return s.Reachability
 }
 
-func (as *AmbientAutoNAT) updateStatus(s NATStatus) {
-	as.status = s
-	switch s {
-	case NATStatusUnknown:
-		as.emitUnknown.Emit(event.EvtLocalRoutabilityUnknown{})
-	case NATStatusPublic:
-		as.emitPublic.Emit(event.EvtLocalRoutabilityPublic{})
-	case NATStatusPrivate:
-		as.emitPrivate.Emit(event.EvtLocalRoutabilityPrivate{})
-	}
+func (as *AmbientAutoNAT) emitStatus() {
+	status := as.status.Load().(autoNATResult)
+	as.emitReachabilityChanged.Emit(event.EvtLocalReachabilityChanged{Reachability: status.Reachability})
 }
 
+// PublicAddr returns the publicly connectable Multiaddr of this node if one is known.
 func (as *AmbientAutoNAT) PublicAddr() (ma.Multiaddr, error) {
-	as.mx.Lock()
-	defer as.mx.Unlock()
-
-	if as.status != NATStatusPublic {
+	s := as.status.Load().(autoNATResult)
+	if s.Reachability != network.ReachabilityPublic {
 		return nil, errors.New("NAT Status is not public")
 	}
 
-	return as.addr, nil
+	return s.address, nil
 }
 
 func (as *AmbientAutoNAT) background() {
 	// wait a bit for the node to come online and establish some connections
 	// before starting autodetection
-	select {
-	case <-time.After(AutoNATBootDelay):
-	case <-as.ctx.Done():
-		return
-	}
+	delay := AutoNATBootDelay
+
+	addrUpdatedChan := as.subAddrUpdated.Out()
+	defer as.subAddrUpdated.Close()
+	defer as.emitReachabilityChanged.Close()
 
 	for {
-		as.autodetect()
+		select {
+		// new connection occured.
+		case conn := <-as.candidatePeers:
+			if conn.Stat().Direction == network.DirInbound &&
+				manet.IsPublicAddr(conn.RemoteMultiaddr()) {
+				ca := as.status.Load().(autoNATResult)
+				if !ca.address.Equal(conn.RemoteMultiaddr()) {
+					as.lastInbound = time.Now()
+				}
+			}
 
-		delay := AutoNATRefreshInterval
-		if as.status == NATStatusUnknown {
-			delay = AutoNATRetryInterval
-		}
+		case <-addrUpdatedChan:
+			if as.confidence > 1 {
+				as.confidence--
```

Comment:
Won't this also trigger when we learn about a _new_ address?

Really, we probably need to call `InterfaceAddrs` on the network every time this event fires, and check to see if our _interface_ addresses have changed. Alternatively, we could just ignore this event and call that every 10 seconds.

Reply:
@Stebalien What's the difference between `Host.Addrs` & `Network. InterfaceAddrs ` ? How does `Network.InterfaceAddrs` affect Reachability ?

Reply:
for reviewing - is it better to keep iterating within this PR, or to have a stabilize branch and structure these sorts of additional logic enhancements as separate PRs into that branch to make them easier to review?

Reply:
Let's get this into a mergable state, merge it, then iterate in future PRs.